README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# caRecall

<!-- badges: start -->

[![R-CMD-check](https://github.com/WraySmith/caRecall/workflows/R-CMD-check/badge.svg)](https://github.com/WraySmith/caRecall/actions)
[![Codecov test
coverage](https://codecov.io/gh/WraySmith/caRecall/branch/main/graph/badge.svg)](https://codecov.io/gh/WraySmith/caRecall?branch=main)
<!-- badges: end -->

The `caRecall` package is an API wrapper for the Government of Canada
[Vehicle Recalls Database
(VRD)](https://tc.api.canada.ca/en/detail?api=VRDB) used by the Defect
Investigations and Recalls Division for vehicles, tires, and child car
seats. The API wrapper provides access to recall summary information
searched using make, model, and year range, as well as detailed recall
information searched using recall number.

The package focuses on querying data from the VRD API to return the
following:

-   Recall summary information
-   Recall counts
-   Detailed recall information

## Installation

The development version of the `caRecall` package can be installed from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("WraySmith/caRecall")
```

## Usage

More documentation on `caRecall` can found at:
<https://wraysmith.github.io/caRecall/>

Note that an API key is required to run the functions in `caRecall` and
query the Vehicle Recalls Database. The key can be acquired at
<https://tc.api.canada.ca/en/detail?api=VRDB>.

A simple example of using the `recall_by_years()` function in the
package to return all recalls manufactured in 2000 could be done as
follows:

``` r
recall_summary_2000 <- recall_by_years(start_year = 2000, end_year = 2000, limit = 3000)
recall_summary_2000
#> # A tibble: 2,422 x 6
#>    `Recall number` `Manufacturer N~ `Model name` `Make name`  Year `Recall date`
#>    <chr>           <chr>            <chr>        <chr>       <int> <date>       
#>  1 1993076         MERCEDES-BENZ    300          MERCEDES-B~  2000 1993-05-31   
#>  2 1999056         FIAT CHRYSLER A~ NEON         CHRYSLER     2000 1999-04-14   
#>  3 1999108         FIAT CHRYSLER A~ NEON         CHRYSLER     2000 1999-06-07   
#>  4 1999111         FLEETWOOD        TIOGA        FLEETWOOD    2000 1999-06-08   
#>  5 1999137         POLARIS          SNOWMOBILE   POLARIS      2000 1999-07-26   
#>  6 1999138         MAZDA            MPV          MAZDA        2000 1999-07-27   
#>  7 1999147         MAZDA            MPV          MAZDA        2000 1999-07-30   
#>  8 1999151         GENERAL MOTORS   S10          CHEVROLET    2000 1999-08-16   
#>  9 1999151         GENERAL MOTORS   SONOMA       GMC          2000 1999-08-16   
#> 10 1999155         GENERAL MOTORS   SUNFIRE      PONTIAC      2000 1999-08-19   
#> # ... with 2,412 more rows
```

The data could then be used to summarize manufacturers with the highest
recall counts in the year:

<img src="man/figures/README-fig1-1.png" width="100%" style="display: block; margin: auto;" />

Additionally, detailed recall information can be queried from the API
using the `recall_details()` function:

``` r
recall_windstar <- recall_details(1997118)
tibble(t(recall_windstar)) #transpose for readability here
#> # A tibble: 15 x 1
#>    `t(recall_windstar)`[,1]                                                     
#>    <chr>                                                                        
#>  1 "1997118"                                                                    
#>  2 "97S69"                                                                      
#>  3 "Light Truck & Van"                                                          
#>  4 "Camionnette et fourgonnette"                                                
#>  5 "WINDSTAR"                                                                   
#>  6 "FORD"                                                                       
#>  7 "27"                                                                         
#>  8 "Seats And Restraints"                                                       
#>  9 "Sièges et dispositifs de retenue"                                           
#> 10 "Safety Mfr"                                                                 
#> 11 "Sécurité - fabricant"                                                       
#> 12 "NOTE: VEHICLES EQUIPPED WITH SECOND ROW BENCH SEATS AND INTEGRATED CHILD SE~
#> 13 "NOTA : Vise les véhicules équipés d’une seconde banquette et du siège intég~
#> 14 "1998"                                                                       
#> 15 "1997-06-16"
```

## Getting Help or Reporting an Issue

To report bugs/issues/feature requests, please file an
[issue](https://github.com/WraySmith/caRecall/issues).

These are very welcome!

## How to Contribute

If you would like to contribute to the package, please see our
[CONTRIBUTING](https://github.com/WraySmith/caRecall/blob/main/CONTRIBUTING.md)
guidelines.

Please note that this project is released with a [Contributor Code of
Conduct](https://github.com/WraySmith/caRecall/blob/main/CODE_OF_CONDUCT.md).
By participating in this project you agree to abide by its terms.


NEWS.md
# caRecall 0.1.0

- First release of the caRecall package.

tests/testthat.R
library(testthat)
library(caRecall)

test_check("caRecall")


tests/testthat/test-utils.R
test_that("API key is retreived", {
  skip_if_no_auth()
  expect_type(get_vrd_key(), "character")
})

test_that("Bad API key correctly fails", {
  # valid url
  url_test <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/make-name/ford"
  query_test <- "ford"
  bad_key <- "test"
  expect_error(call_vrd_api(url_test, query = query_test, api_key = bad_key))
})

test_that("API call returns correct structure", {
  skip_if_no_auth()
  url_test <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/make-name/ford"
  query_test <- "ford"
  response <- call_vrd_api(url_test, query = query_test)
  expect_s3_class(response, "vrd_api", exact = FALSE)
  expect_type(response$response, "list")
  expect_equal(response$response$status_code, 200)
})

test_that("Clean API returns formatted dataframe", {
  skip_if_no_auth()
  # setup for one of the output structures
  # used for all functions except `recall_details()`
  url_test1 <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/make-name/ford"
  query_test1 <- "ford"
  response1 <- call_vrd_api(url_test1, query = query_test1, limit = 25)
  coltype1 <- c(rep("character", 4), "integer", "double")
  # setup for other output structures, used by `recall_detils()`
  url_test2 <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall-summary/recall-number/1977044"
  query_test2 <- "1977045"
  response2 <- call_vrd_api(url_test2, query = query_test2)
  coltype2 <- c(rep("character", 6), "integer", rep("character", 6), "integer", "double")
  # tests for first structure
  expect_s3_class(clean_vrd_api(response1), "data.frame")
  expect_equal(nrow(clean_vrd_api(response1)), 25)
  expect_equal(as.vector(sapply(clean_vrd_api(response1), typeof)), coltype1)
  # tests for second structure
  expect_s3_class(clean_vrd_api(response2), "data.frame")
  expect_equal(nrow(clean_vrd_api(response2)), 1)
  expect_equal(as.vector(sapply(clean_vrd_api(response2), typeof)), coltype2)
})

test_that("Url format is correct", {
  # bad url returns error
  url_with_broken_years <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/make-name/Maz/year-range/2005-2000/count"
  expect_error(check_url(url_with_broken_years))
  # bad url returns error
  url_with_bad_limit <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/model-name/911|ranger/year-range/2008-2100?limit=-10"
  expect_error(check_url(url_with_bad_limit))
  # good urls pass
  url_with_proper_format <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/make-name/Maz/year-range/1995-2000/count"
  expect_equal(check_url(url_with_proper_format), NULL)
  url_with_limit <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/make-name/Maz/year-range/1995-2000/count?limit=10"
  expect_equal(check_url(url_with_limit), NULL)
  url_with_null_limit <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/make-name/Maz/year-range/1995-2000/count?limit="
  expect_equal(check_url(url_with_null_limit), NULL)
  url_with_no_years <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/make-name/Maz/count"
  expect_equal(check_url(url_with_no_years), NULL)
})


tests/testthat/test-recall_data.R
test_that("recall_by_make returns the correct shape", {
  skip_if_no_auth()
  result <- recall_by_make("Nissan", partial = FALSE)
  expect_type(result, "list")
  expect_equal(nrow(result), 25)
  expected_col_names <- c(
    "Recall number", "Manufacturer Name",
    "Model name", "Make name", "Year", "Recall date"
  )
  expect_equal(colnames(result), expected_col_names)
  expected_dtypes <- c(
    "character", "character", "character",
    "character", "integer", "Date"
  )
  returned_dtypes <- as.vector(sapply(result, class))
  expect_equal(returned_dtypes, expected_dtypes)
  names <- unique(result$`Make name`)
  expect_equal(length(names), 1)

  Sys.sleep(1.5)

  result <- recall_by_make("Nissan", start_year = 1980, end_year = 2005)
  expect_gt(nrow(result), 0)
  filtered <- subset(result, Year >= 1980 & Year <= 2005)
  expect_equal(nrow(result), nrow(filtered))
  expected_col_names <- c(
    "Recall number", "Manufacturer Name",
    "Model name", "Make name", "Year", "Recall date"
  )
  expect_equal(colnames(result), expected_col_names)
  expected_dtypes <- c(
    "character", "character", "character",
    "character", "integer", "Date"
  )
  returned_dtypes <- as.vector(sapply(result, class))
  expect_equal(returned_dtypes, expected_dtypes)

  result <- recall_by_make("Nissan", start_year = 1980)
  expect_gt(nrow(result), 0)

  result <- recall_by_make("Nissan", end_year = 1980)
  expect_gt(nrow(result), 0)

  Sys.sleep(1.5)
})

test_that("recall_by_make using partial has the correct shape", {
  skip_if_no_auth()
  result <- recall_by_make("N", partial = TRUE, manufacturer = FALSE)
  names <- unique(result$`Make name`)
  expect_gt(length(names), 1)

  Sys.sleep(1.5)

  result <- recall_by_make("Nissan", partial = FALSE, manufacturer = TRUE)
  names <- unique(result$`Manufacturer Name`)
  expect_equal(length(names), 1)
  expected_col_names <- c(
    "Recall number", "Manufacturer Name",
    "Model name", "Make name", "Year", "Recall date"
  )
  expect_equal(colnames(result), expected_col_names)
  expected_dtypes <- c(
    "character", "character", "character",
    "character", "integer", "Date"
  )
  returned_dtypes <- as.vector(sapply(result, class))
  expect_equal(returned_dtypes, expected_dtypes)

  Sys.sleep(1.5)
})

test_that("recall_by_model has the correct shape", {
  skip_if_no_auth()
  result <- recall_by_model("Civic", limit = 25, partial = TRUE)
  expect_type(result, "list")
  expect_equal(nrow(result), 25)
  expected_col_names <- c(
    "Recall number", "Manufacturer Name",
    "Model name", "Make name", "Year", "Recall date"
  )
  expect_equal(colnames(result), expected_col_names)
  expected_dtypes <- c(
    "character", "character", "character",
    "character", "integer", "Date"
  )
  returned_dtypes <- as.vector(sapply(result, class))
  expect_equal(returned_dtypes, expected_dtypes)
  names <- unique(result$`Model name`)
  expect_gt(length(names), 1)

  Sys.sleep(1.5)

  result <- recall_by_model("Civic", start_year = 1980)
  expect_gt(nrow(result), 0)
  filtered <- subset(result, Year >= 1980)
  expect_equal(nrow(result), nrow(filtered))

  Sys.sleep(1.5)

  result <- recall_by_model("Civic", end_year = 1980)
  expect_gt(nrow(result), 0)
  filtered <- subset(result, Year <= 1980)
  expect_equal(nrow(result), nrow(filtered))

  Sys.sleep(1.5)
})

test_that("recall_by_year returns the correct shape", {
  skip_if_no_auth()
  result <- recall_by_years(2010, 2012)
  expect_type(result, "list")
  expect_equal(nrow(result), 25)
  expected_col_names <- c(
    "Recall number", "Manufacturer Name",
    "Model name", "Make name", "Year", "Recall date"
  )
  expect_equal(colnames(result), expected_col_names)
  filtered <- subset(result, Year >= 2010 & Year <= 2012)
  expect_equal(nrow(result), nrow(filtered))
  expected_dtypes <- c(
    "character", "character", "character",
    "character", "integer", "Date"
  )
  returned_dtypes <- as.vector(sapply(result, class))
  expect_equal(returned_dtypes, expected_dtypes)

  Sys.sleep(1.5)
})

test_that("recall_by_number returns the correct shape", {
  skip_if_no_auth()
  result <- recall_by_number(1977044)
  expect_type(result, "list")
  expect_equal(nrow(result), 1)
  expected_col_names <- c(
    "Recall number", "Manufacturer Name",
    "Model name", "Make name", "Year", "Recall date"
  )
  expect_equal(colnames(result), expected_col_names)
  expected_dtypes <- append(rep("character", 4), c("integer", "Date"))
  returned_dtypes <- as.vector(sapply(result, class))
  expect_equal(returned_dtypes, expected_dtypes)

  Sys.sleep(1.5)
})

test_that("recall_details returns the correct shape when given a list", {
  skip_if_no_auth()
  recall_numbers <- c(1977044, 2009097)
  result <- recall_details(recall_numbers)
  expect_type(result, "list")
  expect_gt(nrow(result), 1)
  expected_col_names <- c(
    "RECALL_NUMBER_NUM", "MANUFACTURER_RECALL_NO_TXT", "CATEGORY_ETXT",
    "CATEGORY_FTXT", "MODEL_NAME_NM", "MAKE_NAME_NM",
    "UNIT_AFFECTED_NBR", "SYSTEM_TYPE_ETXT", "SYSTEM_TYPE_FTXT",
    "NOTIFICATION_TYPE_ETXT", "NOTIFICATION_TYPE_FTXT", "COMMENT_ETXT",
    "COMMENT_FTXT", "DATE_YEAR_CD", "RECALL_DATE_DTE"
  )
  expect_equal(colnames(result), expected_col_names)
  expected_dtypes <- c(
    rep("character", 6),
    c("integer"),
    rep("character", 6),
    c("integer", "Date")
  )
  returned_dtypes <- as.vector(sapply(result, class))
  expect_equal(returned_dtypes, expected_dtypes)

  Sys.sleep(1.5)

  recall_numbers <- seq(1, 700, by = 1)
  expect_error(recall_details(recall_numbers))
})


R/utils.R
#' Retrieves API key from environment
#'
#' Retrieves API key from environment if available.
#'
#' @details
#' An API key is required to use the functions in \code{caRecall} and query the
#' Vehicle Recalls Database. The key can be acquired at
#' \url{https://tc.api.canada.ca/en/detail?api=VRDB}.
#'
#' The key can be set in the environment using
#' \code{Sys.setenv(VRD_API = 'your_API_key_here')}. Alternatively, instead of
#' setting the API key in the environment, the key can be passed into the
#' functions as an argument.
#'
#' @return VRD_API access key if set in environment. If access key is not
#' available, returns an error that the API key is not found and needs to
#' either be set in the environment or passed as an argument into functions.
#' @export
#'
#' @examples
#' \dontrun{
#' get_vrd_key()
#' }
get_vrd_key <- function() {
  x <- Sys.getenv("VRD_API")
  if (x == "") {
    stop("You must provide an API key to the function or use `Sys.setenv(VRD_API = 'your_API_key_here')`")
  }
  x
}

# function used to query the API
# not a user facing function
call_vrd_api <- function(url_, query = NULL, limit = NULL,
                         api_key = NULL) {



  # retrieve API key if no user key input to function call
  if (is.null(api_key)) api_key <- get_vrd_key()

  # add limit to url (limit does not go in header)
  url_ <- paste(url_, "?limit=", toString(limit), sep = "")

  # set the user agent
  ua <- httr::user_agent("https://github.com/WraySmith/caRecall")

  # set headers (currently only user-key is set)
  headers <- httr::add_headers("user-key" = api_key)

  # query the API
  check_url(url_)
  response <- httr::GET(url_, headers, ua)

  ### NEED TO PROVIDE MORE ERROR OUTPUT HERE
  ### POTENTIALLY API KEY AND LINK TO DOCS
  if (httr::status_code(response) != 200) {
    stop(
      sprintf(
        "Vehicle Recall Database request failed [%s]\n%s",
        httr::status_code(response),
        response$request$url
      ),
      call. = FALSE
    )
  }

  # check to verify json response
  if (httr::http_type(response) != "application/json") {
    stop("API did not return json", call. = FALSE)
  }

  # parse response
  parsed <- jsonlite::fromJSON(httr::content(response, "text"), flatten = TRUE)

  # create a class for the returned response
  structure(
    list(
      content = parsed,
      name = query,
      response = response
    ),
    class = "vrd_api"
  )
}

# function used to clean the response from the API call
# not a user facing function
# returns a tibble of output
clean_vrd_api <- function(api_output) {

  # change from json to dataframe format
  unclean_df <- as.data.frame(unlist(api_output$content$ResultSet, recursive = FALSE))

  # checks to verify dataframe is not empty
  if (length(unclean_df) == 0) stop("API query had no results", call. = FALSE)

  # prepares for transpose by removing headers and data type values
  df <- unclean_df[, grep("Literal", colnames(unclean_df))]

  # transposes dataframe
  df <- as.data.frame(t(df), row.names = 1)

  # adds column names
  colnames(df) <- unclean_df$Name

  # checks data type values to change in data frame
  col_type <- unclean_df[, grep("Type", colnames(unclean_df))]

  # check for single row which is not a list
  if (is.list(col_type)) col_type <- col_type[[1]]

  # assigns columns types
  i <- 1
  for (element in col_type) {

    # change df column types
    if (element == "System.Decimal") df[[i]] <- as.integer(df[[i]])
    if (element == "System.DateTime") {
      df[[i]] <- as.Date(
        substr(
          df[[i]],
          1,
          nchar(df[[i]]) - 11
        ),
        "%m/%d/%Y"
      )
    }
    i <- i + 1
  }
  # output as tibble
  tibble::as_tibble(df)
}

# helper function to check for errors in url prior to API call
# not a user facing function
check_url <- function(url_) {

  # check year range
  if (grepl("year-range", url_, fixed = TRUE)) {
    year_range_substring <- gregexpr(pattern = "year-range", url_)
    sub_string <- substring(url_, year_range_substring)
    years_as_string <- unlist(as.list(strsplit(sub_string, "/")[[1]])[2])
    years <- as.list(strsplit(years_as_string, "-")[[1]])
    start_year <- unlist(years[1])
    end_year <- unlist(years[2])
    if (start_year > end_year) {
      stop("Start year must be less than or equal to the end year")
    }
  }

  # test limit
  if (grepl("limit", url_, fixed = TRUE)) {
    limit_substring <- gregexpr(pattern = "limit", url_)
    sub_string <- substring(url_, limit_substring)
    limit_as_string <- unlist(as.list(strsplit(sub_string, "=")[[1]])[2])
    if (!is.null(limit_as_string)) {
      if (as.integer((limit_as_string)) < 1) {
        stop("Limit has to be greater or equal to 1")
      }
    }
  }
  # return NULL if no errors thrown by function
  return(NULL)
}

# helper function to skip unit tests if API key is not available
# not a user facing function
skip_if_no_auth <- function() {
  if (identical(Sys.getenv("VRD_API"), "")) {
    testthat::skip("No authentication available")
  }
}


R/recall_data.R
#' Summary of recalls searching by make or manufacturer
#'
#' Returns summary information of recalls in the Vehicle Recalls Database based
#' on a make or manufacturer search term.
#'
#' @details
#' Queries the Vehicle Recalls Database API by make or manufacturer and returns
#' summary recall information. The year range of the search can be specified
#' and is based on the manufactured year and not the year a recall occurred.
#' Partial search term matches can also be returned by the function and is the
#' default. Note that if \code{partial = FALSE} is used, the number of entries
#' returned by the function may be less than the count provided by
#' \code{\link{count_recall_by_make}} which returns a count for all partial
#' matches.
#'
#' An API key is required to run the function and query the Vehicle Recalls
#' Database. The key can be acquired at
#' \url{https://tc.api.canada.ca/en/detail?api=VRDB}.
#'
#' The API key can be set in the environment using
#' \code{Sys.setenv(VRD_API = 'your_API_key_here')} and will be used by the
#' function, or can be passed into the function using the \code{api_key}
#' argument.
#'
#' @param make List of make or manufacturer names. Case insensitive.
#' @param manufacturer Logical; if TRUE, manufacturer is searched on instead
#' of make.
#' @param start_year Start of year range (optional).
#' @param end_year End of year range (optional).
#' @param limit Number indicating how many recall entries should be returned.
#' Defaults to 25 which is the default of the API.
#' @param partial Logical; if TRUE, returns all partial search term matches.
#' @param api_key API access key to use, if not set in environment.
#'
#' @return A tibble of recall summary information from the Vehicle Recalls
#' Database. Includes six columns.
#' @export
#'
#' @examples
#' \dontrun{
#' recall_by_make("Nissan")
#' recall_by_make(c("Mazda", "Toyota"), start_year = 2008, partial = FALSE)
#' API_KEY <- "xxxxxxxxxxx"
#' recall_by_make("Maz", end_year = 2000, limit = 100, api_key = API_KEY)
#' }
recall_by_make <- function(make, manufacturer = FALSE,
                           start_year = NULL, end_year = NULL,
                           limit = 25, partial = TRUE,
                           api_key = NULL) {

  # create multiple requests from input list
  input_request <- paste(make, collapse = "|")

  # format the url string
  if (manufacturer) {
    url_ <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/manufacturer-name/"
  } else {
    url_ <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/make-name/"
  }

  url_ <- paste(url_, input_request, sep = "")
  # append year range criteria to url if input by user
  if (!is.null(start_year) | !is.null(end_year)) {

    # add start or end year if value one of the values is NULL
    if (is.null(start_year)) start_year <- 1900
    if (is.null(end_year)) end_year <- 2100

    year_range <- paste(toString(start_year), toString(end_year), sep = "-")
    url_ <- paste(url_, "/year-range/", year_range, sep = "")
  }

  # API call, returns class vrd_api
  api_output <- call_vrd_api(url_, make, limit, api_key = api_key)

  # convert content to a tibble
  contents_df <- clean_vrd_api(api_output)

  # filters for exact result
  if (partial == FALSE & manufacturer == FALSE) {
    contents_df <- contents_df[contents_df$`Make name` == toupper(toString(make)), ]
  }
  if (partial == FALSE & manufacturer == TRUE) {
    contents_df <- contents_df[contents_df$`Manufacturer Name` == toupper(toString(make)), ]
  }


  # outputs tibble
  contents_df
}

#' Summary of recalls searching by model
#'
#' Returns summary information of recalls in the Vehicle Recalls Database based
#' on a model search term.
#'
#' @details
#' Queries the Vehicle Recalls Database API by model and returns summary recall
#' information. The year range of the search can be specified and is based on
#' the manufactured year and not the year a recall occurred. Partial search
#' term matches can also be returned by the function and is the default. Note
#' that if \code{partial = FALSE} is used, the number of entries returned by the
#' function may be less than the count provided by
#' \code{\link{count_recall_by_model}} which returns a count for all partial
#' matches.
#'
#' An API key is required to run the function and query the Vehicle Recalls
#' Database. The key can be acquired at
#' \url{https://tc.api.canada.ca/en/detail?api=VRDB}.
#'
#' The API key can be set in the environment using
#' \code{Sys.setenv(VRD_API = 'your_API_key_here')} and will be used by the
#' function, or can be passed into the function using the \code{api_key}
#' argument.
#'
#' @param model List of model names.
#' @param start_year Start of year range (optional).
#' @param end_year End of year range (optional).
#' @param limit Number indicating how many recall entries should be returned.
#' Defaults to 25 which is the default of the API.
#' @param partial Logical; if TRUE, returns all partial search term matches.
#' @param api_key API access key to use, if not set in environment.
#'
#' @return A tibble of recall summary information from the Vehicle Recalls
#' Database. Includes six columns.
#' @export
#'
#' @examples
#' \dontrun{
#' recall_by_model("civic")
#' recall_by_model(c("RANGER RZR 800", "BRZ"), start_year = 2008, partial = FALSE)
#' API_KEY <- "xxxxxxxxxxx"
#' recall_by_model("RANGER", end_year = 2000, limit = 100, api_key = API_KEY)
#' }
recall_by_model <- function(model,
                            start_year = NULL, end_year = NULL,
                            limit = 25, partial = TRUE,
                            api_key = NULL) {

  # create multiple requests from input list
  input_request <- paste(model, collapse = "|")

  # format the url string
  url_ <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/model-name/"
  url_ <- paste(url_, input_request, sep = "")

  # append year range criteria to url if input by user
  if (!is.null(start_year) | !is.null(end_year)) {

    # add start or end year if value one of the values is NULL
    if (is.null(start_year)) start_year <- 1900
    if (is.null(end_year)) end_year <- 2100

    year_range <- paste(toString(start_year), toString(end_year), sep = "-")
    url_ <- paste(url_, "/year-range/", year_range, sep = "")
  }
  # API call, returns class vrd_api
  api_output <- call_vrd_api(url_, model, limit, api_key = api_key)

  # convert content to a tibble
  contents_df <- clean_vrd_api(api_output)

  # filters for exact result
  if (partial == FALSE) {
    contents_df <- contents_df[contents_df$`Model name` == toupper(toString(model)), ]
  }

  # outputs tibble
  contents_df
}

#' Summary of recalls searching by years
#'
#' Returns summary information of recalls in the Vehicle Recalls Database based
#' on a year range search term.
#'
#' @details
#' Queries the Vehicle Recalls Database API by year range and returns
#' summary recall information. If \code{start_year} and/or \code{end_year} are
#' not specified, the function will default to the start and/or end years
#' available in the database. Note that the year range is based on the
#' manufactured year and not the year a recall occurred.
#'
#' An API key is required to run the function and query the Vehicle Recalls
#' Database. The key can be acquired at
#' \url{https://tc.api.canada.ca/en/detail?api=VRDB}.
#'
#' The API key can be set in the environment using
#' \code{Sys.setenv(VRD_API = 'your_API_key_here')} and will be used by the
#' function, or can be passed into the function using the \code{api_key}
#' argument.
#'
#' @param start_year Start of year range.
#' @param end_year End of year range.
#' @param limit Number indicating how many recall entries should be returned.
#' Defaults to 25 which is the default of the API.
#' @param api_key API access key to use, if not set in environment.
#'
#' @return A tibble of recall summary information from the Vehicle Recalls
#' Database. Includes six columns.
#' @export
#'
#' @examples
#' \dontrun{
#' recall_by_years(2010, 2012, limit = 100)
#' API_KEY <- "xxxxxxxxxxx"
#' recall_by_years(end_year = 1970, api_key = API_KEY)
#' }
recall_by_years <- function(start_year = 1900, end_year = 2100,
                            limit = 25, api_key = NULL) {

  # format the url string
  year_range <- paste(toString(start_year), toString(end_year), sep = "-")
  url_ <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/year-range/"
  url_ <- paste(url_, year_range, sep = "")

  # API call, returns class vrd_api
  api_output <- call_vrd_api(url_, year_range, limit,
    api_key = api_key
  )

  # convert content to a tibble
  contents_df <- clean_vrd_api(api_output)


  # output tibble
  contents_df
}

#' Summary of recalls searching by recall number(s)
#'
#' Returns summary information of recalls in the Vehicle Recalls Database based
#' on recall numbers(s).
#'
#' @details
#' Queries the Vehicle Recalls Database API by recall number and returns
#' summary recall information.
#'
#' An API key is required to run the function and query the Vehicle Recalls
#' Database. The key can be acquired at
#' \url{https://tc.api.canada.ca/en/detail?api=VRDB}.
#'
#' The API key can be set in the environment using
#' \code{Sys.setenv(VRD_API = 'your_API_key_here')} and will be used by the
#' function, or can be passed into the function using the \code{api_key}
#' argument.
#'
#' @param recall_number List of recall numbers.
#' @param limit Number indicating how many recall entries should be returned.
#' Defaults to 25 which is the default of the API.
#' @param api_key API access key to use, if not set in environment.
#'
#' @return A tibble of recall summary information from the Vehicle Recalls
#' Database. Includes six columns.
#' @export
#'
#' @examples
#' \dontrun{
#' recall_by_number(1977044)
#' API_KEY <- "xxxxxxxxxxx"
#' recall_by_number(c(2014216, 2013022), api_key = API_KEY)
#' }
recall_by_number <- function(recall_number, limit = 25, api_key = NULL) {

  # create multiple requests from input list
  input_request <- paste(recall_number, collapse = "|")

  # format the url string
  url_ <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall/recall-number/"
  url_ <- paste(url_, input_request, sep = "")

  # API call, returns class vrd_api
  api_output <- call_vrd_api(url_, recall_number, limit,
    api_key = api_key
  )

  # convert content to a tibble
  contents_df <- clean_vrd_api(api_output)

  # output tibble
  contents_df
}

#' Detailed information on recalls searching by recall number(s)
#'
#' Returns detailed information on recalls in the Vehicle Recalls Database
#' based on recall numbers(s).
#'
#' @details
#' Queries the Vehicle Recalls Database API by recall number and returns
#' detailed recall information for each number. Note that each recall number
#' provided to the function results in a separate API call. Accordingly,
#' the rate of API calls is limited to one call per second to stay within the
#' API limit. For example, providing a list of 60 recall numbers to the function
#' would take approximately 1 minute to complete.
#'
#' An API key is required to run the function and query the Vehicle Recalls
#' Database. The key can be acquired at
#' \url{https://tc.api.canada.ca/en/detail?api=VRDB}.
#'
#' The API key can be set in the environment using
#' \code{Sys.setenv(VRD_API = 'your_API_key_here')} and will be used by the
#' function, or can be passed into the function using the \code{api_key}
#' argument.
#'
#' @param recall_number List of recall numbers.
#' @param api_key API access key to use, if not set in environment.
#'
#' @return A tibble of detailed recall information from the Vehicle Recalls
#' Database. Includes 15 columns.
#' @export
#'
#' @examples
#' \dontrun{
#' recall_details(1977044)
#' API_KEY <- "xxxxxxxxxxx"
#' recall_details(c(2014216, 2013022), api_key = API_KEY)
#' }
recall_details <- function(recall_number, api_key = NULL) {

  # a limit of 60 calls are allowed per minute
  # this function rate limits and a call with 600 numbers would take > 10 min
  if (length(recall_number) > 600) {
    stop(
      print("Number of recall_numbers must be less than 600"),
      call. = FALSE
    )
  }


  i <- 1
  for (single_number in recall_number) {

    # format the url string
    url_ <- "https://vrdb-tc-apicast-production.api.canada.ca/eng/vehicle-recall-database/v1/recall-summary/recall-number/"
    url_ <- paste(url_, toString(single_number), sep = "")

    # API call, returns class vrd_api
    api_output <- call_vrd_api(url_, single_number,
      api_key = api_key
    )

    if (i == 1) {
      # initialize tibble
      compiled_df <- clean_vrd_api(api_output)
    } else {
      # append to tibble
      compiled_df <- rbind(compiled_df, clean_vrd_api(api_output))
    }

    Sys.sleep(1)
    i <- i + 1
  }
  # output tibble
  compiled_df
}
